tracks/javascript/exercises/linked-list/mentoring.md

```diff
@@ -0,0 +1,160 @@
+      this.head = this.head.next
+      this.head.prev = null
+    } else {
+      this.clear()
+    }
+
+    return value
+  }
+
+  shift() {
+    if (!this.tail) {
+      throw new Error('list is empty')
+    }
+
+    const { value } = this.tail
+
+    if (this.tail.prev) {
+      this.tail = this.tail.prev
+      this.tail.next = null
+    } else {
+      this.clear()
+    }
+
+    return value;
+  }
+
+  count() {
+    let count = 0
+    let element = this.head
+
+    while (element) {
+      count += 1
+      element = element.next
+    }
+
+    return count
+  }
+
+  delete(value) {
+    let element = this.head
+    while (element) {
+      if (element.value !== value) {
+        element = element.next
+        // eslint-disable-next-line no-continue
+        continue
+      }
+
+      if (element.next) {
+        element.next.prev = element.prev
+      } else {
+        this.tail = this.tail.prev
+      }
+
+      if (element.prev) {
+        element.prev.next = element.next
+      } else {
+        this.head = this.head.next
+      }
+
+      break
+    }
+  }
+}
+```
+
+Variations include the following
+- A student may only pass `value` to `Node` and assign `prev` and `next` afterwards
+- A student may forego the `throw` clauses completely as there are no test cases covering empty list popping and shifting.
+- A student may actually put the logic on the `Node` class (classic linked list) and only have the `LinkedList` API keep track of `head` and `tail` (and fix them on mutations). `LinkedList` is than a mere proxy to the actual doubly linked-list.
+- A student may traverse from `tail` backwards instead of `head` forwards.
+- A student may use a `NO_ELEMENT` element instead of `null` or `undefined` to indicate there is no element at `next` / `head` / `prev` / `tail`.
+
+#### Note about Array solutions
+The current tests are _not_ enforcing this implementation to actually be a
+LinkedList, but the instructions clearly say you'll need to create a `Node` or
+`Element` data structure pointing to `next` and `prev`. The following _could_ be
+accepted but you should encourage the student to actually implement this as a
+linked list:
+
+```javascript
+export class LinkedList extends Array { }
+
+LinkedList.prototype.count = function() {
+  return this.length;
+}
+
+LinkedList.prototype.delete = function(value) {
+  const index = this.indexOf(value)
+  if (index !== -1) {
+    this.splice(index, 1)
+  }
+}
+```
+
+Variations on this include array destructuring in the implementation of `pop`,
+`push` etc, or using a backing array inside the `LinkedList` class. See *common
+suggestions* for more information.
+
+#### Note about earlier versions
+Earlier version had a different `Readme.MD` that pointed to using a `Deque`.
+This usually only confused students and has been removed.
+
+### Common suggestions
+- If a student uses a backing array or the likes, talk about performance and the upsides of actually using a linked-list.
```

Comment:
Also talk about `linked lists` being [classical data structures](https://en.wikipedia.org/wiki/Linked_data_structure) and it is a good exercise to implement them in the intended way.